apps/meteor/app/livechat/server/lib/contacts/ContactMerger.ts

```diff
@@ -0,0 +1,307 @@
+	ILivechatVisitor,
+	ILivechatContactChannel,
+	ILivechatContactConflictingField,
+	IUser,
+	DeepWritable,
+	IOmnichannelSource,
+} from '@rocket.chat/core-typings';
+import { LivechatContacts } from '@rocket.chat/models';
+import type { ClientSession, UpdateFilter } from 'mongodb';
+
+import { getContactManagerIdByUsername } from './getContactManagerIdByUsername';
+import { isSameChannel } from '../../../lib/isSameChannel';
+
+type ManagerValue = { id: string } | { username: string };
+type ContactFields = {
+	email: string;
+	phone: string;
+	name: string;
+	username: string;
+	manager: ManagerValue;
+	channel: ILivechatContactChannel;
+};
+
+type CustomFieldAndValue = { type: `customFields.${string}`; value: string };
+
+export type FieldAndValue =
+	| { type: keyof Omit<ContactFields, 'manager' | 'channel'>; value: string }
+	| { type: 'manager'; value: ManagerValue }
+	| { type: 'channel'; value: ILivechatContactChannel }
+	| CustomFieldAndValue;
+
+type ConflictHandlingMode = 'conflict' | 'overwrite' | 'ignore';
+
+type MergeFieldsIntoContactParams = {
+	fields: FieldAndValue[];
+	contact: ILivechatContact;
+	conflictHandlingMode?: ConflictHandlingMode;
+	session?: ClientSession;
+};
+
+export class ContactMerger {
+	private managerList = new Map<IUser['username'], IUser['_id'] | undefined>();
+
+	private getManagerId(manager: ManagerValue): IUser['_id'] | undefined {
+		if ('id' in manager) {
+			return manager.id;
+		}
+
+		return this.managerList.get(manager.username);
+	}
+
+	private isSameManager(manager1: ManagerValue, manager2: ManagerValue): boolean {
+		if ('id' in manager1 && 'id' in manager2) {
+			return manager1.id === manager2.id;
+		}
+		if ('username' in manager1 && 'username' in manager2) {
+			return manager1.username === manager2.username;
+		}
+
+		const id1 = this.getManagerId(manager1);
+		const id2 = this.getManagerId(manager2);
+
+		if (!id1 || !id2) {
+			return false;
+		}
+
+		return id1 === id2;
+	}
+
+	private isSameField(field1: FieldAndValue, field2: FieldAndValue): boolean {
+		if (field1.type === 'manager' && field2.type === 'manager') {
+			return this.isSameManager(field1.value, field2.value);
+		}
+
+		if (field1.type === 'channel' && field2.type === 'channel') {
+			return isSameChannel(field1.value.visitor, field2.value.visitor);
+		}
+
+		if (field1.type !== field2.type) {
+			return false;
+		}
+
+		if (field1.value === field2.value) {
+			return true;
+		}
+
+		return false;
+	}
+
+	private async loadDataForFields(session?: ClientSession, ...fieldLists: FieldAndValue[][]): Promise<void> {
+		for await (const fieldList of fieldLists) {
+			for await (const field of fieldList) {
+				if (field.type !== 'manager' || 'id' in field.value) {
+					continue;
+				}
+
+				if (this.managerList.has(field.value.username)) {
+					continue;
+				}
+
+				const id = await getContactManagerIdByUsername(field.value.username, session);
+				this.managerList.set(field.value.username, id);
+			}
+		}
+	}
```

Comment:
```suggestion
	private static async createWithFields(session?: ClientSession, ...fieldLists: FieldAndValue[][]) {
		const merger = new ContactMerger();
		for await (const fieldList of fieldLists) {
			for await (const field of fieldList) {
				if (field.type !== 'manager' || 'id' in field.value) {
					continue;
				}

				if (merger.managerList.has(field.value.username)) {
					continue;
				}

				const id = await getContactManagerIdByUsername(field.value.username, session);
				merger.managerList.set(field.value.username, id);
			}
		}
		return merger;
	}
```

Reply:
The non-static functions were used for the manager cache